File: frontend/src/app/modules/customer/services/customer.service.ts
import {Injectable} from '@angular/core';
import {Observable} from "rxjs";
import {ApiService} from 'src/app/shared/services/api/api.service';
import {CustomerModel} from '../models/customer.model';
import {ApiResponseModel} from 'src/app/shared/models/api/api-response.model';
import {ApiEndpointConstant} from 'src/app/shared/constants/api-endpoint.constant';
import {HttpParams} from "@angular/common/http";
import {PaginationRequestModel} from "../../../shared/models/pagination/pagination-request.model";

@Injectable({
  providedIn: 'root'
})
export class CustomerService  extends ApiService {


  public getCustomers(pagination: PaginationRequestModel): Observable<ApiResponseModel<any>> {
    return this.get<ApiResponseModel<any>>(ApiEndpointConstant.CUSTOMER, new HttpParams({fromObject: pagination as any}));
  }

  public getCustomer(username: string): Observable<ApiResponseModel<any>> {
    return this.get<ApiResponseModel<any>>(`${ApiEndpointConstant.CUSTOMER}/${username}`);
  }

  public getCustomerCount(): Observable<ApiResponseModel<any>> {
    return this.get<ApiResponseModel<any>>(`${ApiEndpointConstant.CUSTOMER}${ApiEndpointConstant.COUNT}`);
  }

  public registerCustomer(customerModel: CustomerModel): Observable<ApiResponseModel<any>> {
    return this.post<ApiResponseModel<any>>(ApiEndpointConstant.CUSTOMER_REGISTER, customerModel, {
      withCredentials: true
    });
  }

}

File: backend/src/main/java/com/tasklion/backend/features/tasklionUser/customer/service/CustomerService.java
package com.tasklion.backend.features.tasklionUser.customer.service;

import com.tasklion.backend.common.model.PaginationModel;
import com.tasklion.backend.features.auth.model.AuthResponseModel;
import com.tasklion.backend.features.tasklionUser.customer.Customer;
import com.tasklion.backend.features.tasklionUser.customer.CustomerModel;
import org.springframework.data.domain.Page;

public interface CustomerService {

    Page<CustomerModel> getCustomers(PaginationModel paginationModel);

    CustomerModel getCustomer(String username);

    Customer saveCustomer(CustomerModel customerModel);

    long getCustomerCount();

    AuthResponseModel registerCustomer(CustomerModel customerModel);

}
